Document PDF export (#2706)
telerik/blazor-docs: components/grid/export/pdf.md

> # PDF Export

Reviewer: We use the component name in article titles for SEO. The articles in the Export section doesn't seem to follow this, so consider updating if you like.

> You can export the Grid to PDF with a click of a button. The current filter, sort, page, grouping, column order and column size are applied to the exported PDF document.

Reviewer 2: `a click` indicates a random, unidentified click, `the` is more definitive. Note that the [Style Guide](https://docs.telerik.com/style-guide/punctuation#series) calls for the use of the serial comma (which is counter-intuitive for some other languages).

```suggestion
You can export the Grid to PDF with the click of a button. The current filter, sort, page, grouping, column order, and column size are applied to the exported PDF document.
```

When you click the Export button, your browser will receive the resulting file.

#### In This Article

>   - [Basics](#basics)

Reviewer 2: You don't really need the indentation here.

Author: This follows the model from the other articles. We include such indentation in every "In This Article" section - e.g. 
https://www.telerik.com/blazor-ui/documentation/components/stepper/steps/indicators

  - [Programmatic Export](#programmatic-export)
  - [Customization](#customization)
  - [Notes](#notes)
  - [Custom Export](#custom-export)

## Basics

To enable the Grid PDF Export, add a [command button](slug:components/grid/columns/command) with the `PdfExport` command name to the [Grid toolbar](slug:components/grid/features/toolbar).

````RAZOR.skip-repl

> <GridToolBarTemplate>

Reviewer: You can also show how to do this with the new toolbar syntax.

Author: In an offline discussion we've agreed to do that at a later stage for all examples: https://github.com/telerik/blazor/issues/10894

    <GridCommandButton Command="PdfExport" Icon="@SvgIcon.FilePdf">Export to Pdf</GridCommandButton>
</GridToolBarTemplate>
````

Optionally, you can also set the `GridPdfExport` tag settings under the `GridExport` tag to subscribe to the [Grid export events](slug:grid-export-events) that allow further customization of the exported columns/data or configure the PDF export options:

@[template](/_contentTemplates/common/parameters-table-styles.md#table-layout)

| Parameter | Type and Default&nbsp;Value | Description |
| --- | --- | --- |
| `FileName` | `string` | The name of the file. The Grid will add the `.pdf` extension for you. |
| `AllPages` | `bool` |  Whether to export the current page only, or the entire data from the data source. |
| `PaperSize` | `GridPdfExportPaperSize` enum <br/> (`A4`) | The size of the paper for the exported file. |

> | `PageOrientation` | `GridPdfExportPageOrientation` enum <br/> (`Portrait`)| The orientation of the page - portrait and landscape. |

Reviewer 2: ```suggestion
| `PageOrientation` | `GridPdfExportPageOrientation` enum <br/> (`Portrait`)| The orientation of the page&mdash;portrait or landscape. |
```

> >caption Export the Grid to PDF - Example

Reviewer: ```suggestion
>caption Export the Grid to PDF
```

@[template](/_contentTemplates/common/parameters-table-styles.md#table-layout)

| Method | Type | Description |
| --- | --- | --- |
| `ExportToPdfAsync` | `Task<MemoryStream>` | Returns the exported data as a `MemoryStream`. The stream itself is finalized, so that the resource does not leak. To read and work with the stream, clone its available binary data to a new `MemoryStream` instance. |
| `SaveAsPdfFileAsync` | `ValueTask` | Sends the exported PDF file to the browser for download. |

>caption Invoke the PDF export function from code

````RAZOR
@* Send the exported file for download and get the exported data as a memory stream *@

@using System.IO

<TelerikButton OnClick="@(async () => await GridRef.SaveAsPdfFileAsync())">Download the PDF file</TelerikButton>
<TelerikButton OnClick="@GetTheDataAsAStream">Get the Exported Data as a MemoryStream</TelerikButton>

<TelerikGrid @ref="@GridRef"
             Data="@GridData"
             FilterMode="@GridFilterMode.FilterMenu"
             Pageable="true"
             Sortable="true"
             Reorderable="true"
             Resizable="true"
             Groupable="true"

>              Width="650px">

Reviewer: Consider alphabetic sorting with so many parameters.

        <GridColumn Field="@nameof(SampleData.ProductId)" Title="ID" Width="100px" />
        <GridColumn Field="@nameof(SampleData.ProductName)" Title="Product Name" Width="200px" />
        <GridColumn Field="@nameof(SampleData.UnitsInStock)" Title="In stock" Width="100px" />
        <GridColumn Field="@nameof(SampleData.Price)" Title="Unit Price" Width="100px" />
        <GridColumn Field="@nameof(SampleData.Discontinued)" Title="Discontinued" Width="150px" />
    </GridColumns>
</TelerikGrid>

@code {
    private TelerikGrid<SampleData> GridRef { get; set; }

    private MemoryStream exportedPdfStream { get; set; }

    private List<SampleData> GridData { get; set; }

    private bool ExportAllPages { get; set; }

    private async Task GetTheDataAsAStream()
    {
        MemoryStream finalizedStream = await GridRef.ExportToPdfAsync();

        exportedPdfStream = new MemoryStream(finalizedStream.ToArray());
    }

    protected override void OnInitialized()
    {
        GridData = Enumerable.Range(1, 100).Select(x => new SampleData
            {
                ProductId = x,
                ProductName = $"Product {x}",
                UnitsInStock = x * 2,

>                 Price = 3.14159m * x,

Reviewer: Why so much precision? You can also use `Random.Shared.Next()`

Author: Random.Shared.Next( ) returns an integer, while this is a more "realistic" price value. We can indeed reduce the precision a bit, though, will do.

                Discontinued = x % 4 == 0,
            }).ToList();
    }

    public class SampleData
    {
        public int ProductId { get; set; }
        public string ProductName { get; set; }
        public int UnitsInStock { get; set; }
        public decimal Price { get; set; }
        public bool Discontinued { get; set; }
    }
}
````

## Customization

To customize the exported file, handle the `OnBeforeExport` or `OnAfterExport` events the Grid exposes. 

The component allows you to control the data set that will be exported. It also provides built-in customization options for the columns such as `Width`, `Title` and more.

> For more advanced customization the Grid lets you get the `MemoryStream` of the file. Thus, you can customize it using the [`PdfProcessing`](https://docs.telerik.com/devtools/document-processing/libraries/radpdfprocessing/overview) library that is available with your license.

Reviewer: It is

Read more about how to [customize the exported file](slug:grid-export-events).

> ## Notes

Reviewer: Notes sections tend to contains unstructured and unfindable information, and don't scale well at all. Consider:

* Move some of these to the Basics section, for example the info about column widths, as it's important.
* User H3 headings in Notes or somewhere else to provide granular and linkable structured information.

Author: Will be handled as part of https://github.com/telerik/blazor/issues/10895.

* PDF does not understand units different than `px` for the column `Width`, and if you use them (such as `rem` or `%`), it will fail to parse them and will render a collapsed (hidden) column with zero width.
* When exporting Grid columns, you must provide appropriate `PaperSize` and `PageOrientation` properties. For example, if you want to render 20 columns (100px each) in an A4 sheet, then this will yield unexpected results. The column dimensions in a PDF file are fixed, thus they cannot be resized as in Excel, which requires the developer to ensure proper export dimensions.
* Exporting to PDF in UI for Blazor is different from exporting in Kendo jQuery, where the full HTML is exported. The Blazor export to PDF will export the Grid to a table, similar to an Excel table. If you want [to export to PDF as HTML, you can use a custom approach](#custom-export).

> * Templates are not exported, because there is no provision in the framework for getting them at runtime. If a column, header or group header/footer has a template or aggregates, it will be ignored. The headers will be the `Title` of the column, the data is the data from the `Field`. If you need additional information, see if you can add it in a Field in the model, or create your own PDF file. Find a <a href="https://feedback.telerik.com/blazor/1485764-customize-the-Pdf-file-before-it-gets-to-the-client" target="_blank">project example on how to generate your own exported file</a>. Find additional information on how to [export an image that is rendered in a Grid column template](slug:grid-export-image-column-Pdf).

Reviewer: This section should lead to the Custom Export. Even better, you can separate this template information together with the Custom Export information in a single section with a heading of your preference.

Author: Will be handled as part of https://github.com/telerik/blazor/issues/10895.

> * Templates are not exported, because there is no provision in the framework for getting them at runtime. If a column, header or group header/footer has a template or aggregates, it will be ignored. The headers will be the `Title` of the column, the data is the data from the `Field`. If you need additional information, see if you can add it in a Field in the model, or create your own PDF file. Find a <a href="https://feedback.telerik.com/blazor/1485764-customize-the-Pdf-file-before-it-gets-to-the-client" target="_blank">project example on how to generate your own exported file</a>. Find additional information on how to [export an image that is rendered in a Grid column template](slug:grid-export-image-column-Pdf).

Author: Will be handled as part of https://github.com/telerik/blazor/issues/10895.

> @[template](/_contentTemplates/grid/export.md#export-common-notes)

Reviewer 3: I propose here, updating the example to include the PDF Export option. Also, you can polish the example and make it runnable.

Author: Will be handled as part of https://github.com/telerik/blazor/issues/10895.
